torch/csrc/jit/codegen/cuda/root_domain_map.h

```diff
@@ -0,0 +1,331 @@
+#pragma once
+
+#include <torch/csrc/jit/codegen/cuda/disjoint_set.h>
+#include <torch/csrc/jit/codegen/cuda/iter_visitor.h>
+#include <torch/csrc/jit/codegen/cuda/utils.h>
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+class TORCH_CUDA_API RootDomainMap : public PolymorphicBase {
+ public:
+  //! Return a map from a producer TensorDomain to a consumer
+  //! TensorDomain
+  //!
+  //! \param producer A producer TensorDomain
+  //! \param consumer A consumer TensorDomain
+  //! \param root_dims_to_map Maps only producer root domains in this set
+  std::unordered_map<IterDomain*, IterDomain*> mapProducerToConsumer(
+      const TensorDomain* producer,
+      const TensorDomain* consumer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map) const;
+
+  //! Return a map from a consumer TensorDomain to a producer
+  //! TensorDomain
+  //!
+  //! \param consumer A consumer TensorDomain
+  //! \param producer A producer TensorDomain
+  //! \param root_dims_to_map Maps only consumer root domains in this set
+  std::unordered_map<IterDomain*, IterDomain*> mapConsumerToProducer(
+      const TensorDomain* consumer,
+      const TensorDomain* producer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map) const;
+
+  virtual std::ostream& print(std::ostream& os) const = 0;
+
+ protected:
+  //! Return a map between root IterDomains of a producer-consumer
+  //! pair.
+  //!
+  //! \param producer A producer TensorDomain
+  //! \param consumer A consumer TensorDomain
+  //! \param root_dims_to_map Maps only from IterDomains in this set
+  //! \param producer_to_consumer Maps from producer to consumer if true
+  virtual std::unordered_map<IterDomain*, IterDomain*> map(
+      const TensorDomain* producer,
+      const TensorDomain* consumer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map,
+      bool producer_to_consumer) const = 0;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const RootDomainMap& map) {
+  return map.print(os);
+}
+
+class TORCH_CUDA_API PairwiseRootDomainMap : public RootDomainMap {
+ public:
+  explicit PairwiseRootDomainMap(
+      const TensorView* producer,
+      const TensorView* consumer);
+
+  std::ostream& print(std::ostream& os) const override;
+
+ protected:
+  std::unordered_map<IterDomain*, IterDomain*> map(
+      const TensorDomain* producer,
+      const TensorDomain* consumer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map,
+      bool producer_to_consumer) const override;
+
+  //! Creates mapping that does not consider the actual expression of
+  //! a producer-consumer. This is potentially unsafe.
+  PairwiseRootDomainMap() = default;
```

Comment:
Do you mean doesn't consider all expressions in the fusion? The constructor is looking at the expression containing consumer and producer.

Reply:
Sorry, the comment is misleading. When `PairwiseRootDomainMap` is created with this interface, its `producer_` and `consuemr_` TensorViews are null, so the mapping won't be able to use the bcast flags, thus it's potentially unsafe. Since we don't need to have that unsafe interface, this constructor is going to be removed.